### DISCUSSION TITLE
Dag with @monthly schedule triggers before complete period ends

### DISCUSSION THREAD

**Discussion by vxtals** [apache/airflow#21131] [Q&A] [answered]
### Apache Airflow version

2.2.2

### What happened

Today, 2022-01-26, I created a dag with start_date '2021-12-01 11:09:00' and a schedule_interval '@monthly'. I understand this is expected to execute for the first time on february 1st, because at '2022-01-01 0:00:00', the month period is not completed yet. But it started as soon as I turned it on. 

Is this ok?  

### What you expected to happen

Scheduled execution shouldn't happen until february 1st, when there is a completed period.

### How to reproduce

[test_trigger.txt](https://github.com/apache/airflow/files/7942850/test_trigger.txt)

Change this Dag file extensión to .py and add it to your dags folder.

### Operating System

macos locally and redhat ubi-minimal in production

### Deployment

Other Docker-based deployment

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Your start date is in december. The first completed period ended 1.01.2021 and that was your first run for.

**Reply by vxtals** [reply]
If start date is 2021-12-01 11:09:00, then on 2022-01-01 00:00:00 there is not a complete month yet, there are 11 hours and 9 minutes missing. And this used to be the behavior in all versions up to 2.0.1. 

We had airflow 2.0.1 and a Dag with this schedule and start date, it was turned on and it was manually executed on january 3rd. When we upgrade yesterday to 2.2.2,  the execution for december period was suddenly triggered. Why it wasn't triggered before on 2.0.1 if it had same schedule and start date and it was on?

Thank you.

**Reply by potiuk** [reply]
I don't know - but it's likely some edge case due to changes in customizable timetables.

Out of curiosity what makes you think that the "full period shoudl be there" ? Is it documented somewhere? Where? 

It does not seem obvious and might be side effect and your interpretation of the behaviour, but I do not recall it has been specified somewhere in Airflow. It does not seem obvious as a feature. 

But I might be wrong, and not know something, so maybe you can point to such docs/descriiption where you got the impression it should work this way? (Even if it worked like that before, it could be just unintended side effect rather than feature).

**Reply by vxtals** [reply]
Here in the docs:

https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html

> @monthly | Run once a month at midnight of the first day of the month 

> Similarly, since the start_date argument for the DAG and its tasks points to the same logical date, it marks the start of the DAG’s first data interval, not when tasks in the DAG will start running. In other words, a DAG run will only be scheduled one interval after start_date.

With the last phrase I understand that one interval must be completed before first execution.

**Reply by potiuk** [reply]
I tink it depends on the interpretation, really. I was not very close to that part, so I am not entirely sure what was the intention but maybe other Airflow maintainers from the past might shed some light on it :).

My best guess is that this is a very subtle change induced by the timetable introduction @uranusjr @ashb - I think the behaviour is diferent and it seems it *might* be result of  the difference in the implementations of custom timetables vs. past CRON behaviour..

Clearly it looks like a behvioural change of Airflow, but I think the question is:

a) was it intentional
b) was it really "bad" change

This seems quite lilkely - unlike the past CRON implementation - the current timtable implementaiton does not assume that the interval has "fixed" length  so it might simply not look in the past whether the previous interval was "fully completed" or not. 

WDYT?

**Reply by uranusjr** [reply]
Actually there’s a PR for this: #21011, and I reviewed it even _before this conversation existed_ assuming it’s a bug. That makes two people being confused, so let’s say it’s bug and fix it?

**Reply by potiuk** [reply]
Yeah. That was the discussion you referred to - yeah. Let's fix it.

**Comment by ashb** [comment, marked as answer]
`@monthly` plus a non-round start date are likely the culprit here, and the behaviour is/was incompletely specified.

The start of the next `@monthly` period after 2021-12-01T11:09 is 2022-01-01T00:00:00, so the first month is 11 hours shorter, but the second and subsequent months end at midnight UTC of the 1st of the next month.

You can see this behaviour with `airflow dags next-execution` command.

From what you've described this sounds like a bug in the old implementation to me.

I think (but I haven't checked) that `@daily` on 2.0.x and 2.2.x would give this same behavoiur -- first run would be 11:09->Midnight, but subsequent runs would be Midnight->Midnight, as `@daily` is shortcut for `0 0 * * *` and monthly is shortcut for `0 0 1 * * https://crontab.guru/#@monthly

**Reply by potiuk** [reply]
Yeah. That would be my understanding too.

**Reply by potiuk** [reply]
Cool!

**Comment by vxtals** [comment]
I guess I saw that behavior and assumed it was expected. Thank you anyway.
